parchmint/similaritymatcher.py

```diff
@@ -0,0 +1,131 @@
+        self.G1_device = G1
+        self.G2_device = G2
+        self.compare_params = compare_params
+        self.G1_param_diff_list = []
+        self.G2_param_diff_list = []
+        self.G1_layer_diff_list = []
+        self.G2_layer_diff_list = []
+        self.G1_port_diff_list = []
+        self.G2_port_diff_list = []
+
+        super(SimilarityMatcher, self).__init__(G1.G, G2.G)
+
+    def semantic_feasibility(self, G1_node, G2_node) -> bool:
+        """Returns True if adding (G1_node, G2_node) is symantically feasible.
+        The semantic feasibility function should return True if it is
+        acceptable to add the candidate pair (G1_node, G2_node) to the current
+        partial isomorphism mapping.   The logic should focus on semantic
+        information contained in the edge data or a formalized node class.
+        By acceptable, we mean that the subsequent mapping can still become a
+        complete isomorphism mapping.  Thus, if adding the candidate pair
+        definitely makes it so that the subsequent mapping cannot become a
+        complete isomorphism mapping, then this function must return False.
+        The default semantic feasibility function always returns True. The
+        effect is that semantics are not considered in the matching of G1
+        and G2.
+        The semantic checks might differ based on the what type of test is
+        being performed.  A keyword description of the test is stored in
+        self.test.  Here is a quick description of the currently implemented
+        tests::
+          test='graph'
+            Indicates that the graph matcher is looking for a graph-graph
+            isomorphism.
+          test='subgraph'
+            Indicates that the graph matcher is looking for a subgraph-graph
+            isomorphism such that a subgraph of G1 is isomorphic to G2.
+        Any subclass which redefines semantic_feasibility() must maintain
+        the above form to keep the match() method functional. Implementations
+        should consider multigraphs.
+        """
+
+        # check each components. If not same, print out the difference and return false. 
+        feasible = True
+
+
+        G1_component = self.G1_device.get_component(G1_node)
+        G2_component = self.G2_device.get_component(G2_node)
+
+
+        if G1_component.layers != G2_component.layers:
+          print("layer wrong")
+          feasible = False
+        
+        # store all the differences
+
+        if G1_component.params != G2_component.params:
+          print("params wrong")
+          self.G1_param_diff_list.append(G1_component.params)
+          self.G2_param_diff_list.append(G2_component.params)
+
+          if self.compare_params is True:
+            feasible = False
+          else:
+            feasible = True          
+
+        if G1_component.ports != G2_component.ports:
+          print("ports wrong")
+          feasible = False
+
```

Comment:
You should also find all the incoming/outgoing connections to these components and make sure they're connecting the same ports. 

```
MIXER m1;
MIXER m2;
CHANNEL c1 from m1 1 to m2 1;
```
This is different from:
```
MIXER m1;
MIXER m2;
CHANNEL c1 from m1 1 to m2 2;
```
Add another argument called `check_connection_targets` and do the same logic as `check_params`.